drivers/ccs811/ccs811.c

```diff
@@ -0,0 +1,684 @@
+/*
+ * Copyright (C) 2018 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ccs811
+ * @brief       Device Driver for AMS CCS811 digital gas sensor
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @file
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "log.h"
+#include "xtimer.h"
+
+#include "ccs811_regs.h"
+#include "ccs811.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * Internal macro definitions
+ */
+
+#define ASSERT_PARAM(cond) \
+    if (!(cond)) { \
+        DEBUG("[ccs811] %s: %s\n", \
+              __func__, "parameter condition (" # cond ") not fulfilled"); \
+        assert(cond); \
+    }
+
+#define DEBUG_DEV(f, d, ...) \
+    DEBUG("[ccs811] %s dev=%d addr=%02x: " f "\n", \
+          __func__, d->params.i2c_dev, d->params.i2c_addr, ## __VA_ARGS__)
+
+#define ERROR_DEV(f, d, ...) \
+    LOG_ERROR("[ccs811] dev=%d addr=%x: " f "\n", \
+              d->params.i2c_dev, d->params.i2c_addr, ## __VA_ARGS__)
+
+/**
+ * Internal type declarations
+ */
+
+typedef struct {
+    uint8_t reserved_1 : 2;
+    uint8_t int_thresh : 1; /**< interrupt if new ALG_RESULT_DAT crosses on of the thresholds */
+    uint8_t int_datardy: 1; /**< interrupt if new sample is ready in ALG_RESULT_DAT  */
+    uint8_t drive_mode : 3; /**< mode number binary coded */
+} ccs811_meas_mode_reg_t;
+
+/**
+ * forward declaration of functions for internal use only
+ */
+static int _reg_read(const ccs811_t *dev, uint8_t reg, uint8_t *data, uint32_t len);
+static int _reg_write(const ccs811_t *dev, uint8_t reg, uint8_t *data, uint32_t len);
+static int _check_error_status(const ccs811_t *dev);
+static int _error_code(const ccs811_t *dev, uint8_t err_reg);
+static int _is_available(const ccs811_t *dev);
+
+int ccs811_init(ccs811_t *dev, const ccs811_params_t *params)
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(params != NULL);
+
+    /* init sensor data structure */
+    dev->params = *params;
+
+    int res = CCS811_OK;
+
+#if MODULE_CCS811_FULL
+    if (dev->params.reset_pin != GPIO_UNDEF &&
+        gpio_init(dev->params.reset_pin, GPIO_OUT) == 0) {
+        DEBUG_DEV("nRESET pin configured", dev);
+        /* enable low active reset signal */
+        gpio_clear(dev->params.reset_pin);
+        /* t_RESET (reset impuls) has to be at least 20 us, we wait 1 ms */
+        xtimer_usleep(1000);
+        /* disable low active reset signal */
+        gpio_set(dev->params.reset_pin);
+        /* t_START after reset is 1 ms, we wait 1 further ms */
+        xtimer_usleep(1000);
+    }
+    else {
+        dev->params.reset_pin = GPIO_UNDEF;
+        DEBUG_DEV("nRESET pin not configured or could not be used", dev);
+    }
+
+    if (dev->params.wake_pin != GPIO_UNDEF &&
```

Comment:
The wake_pin is mandatory for the basic version of the driver as well. Otherwise the basic test driver doesn't work: the device is not reachable and cannot be initialized.

Reply:
I'm not sure. We define the base version

- to save memory and
- to keep the number of required GPIOs small.

Pulling down the WAKE pin permanently is a valid option which is also mentioned in the datasheet: "If power is not an issue, nWake can be connected to ground if a spare GPIO is not available on the host processor."

Ok, we could say, the code size for handling the wakeup pin is small in comparison to the remaining code and if the user has no GPIOs left, he has the option to set wakeup pin to GPIO_UNDEF in the parameter definition.

I'm just want to be sure since, the last approach would require some changes.

Reply:
I measured the code size for Arduino MKR1000 (samd21).

module version | code size of ccs811.a
---------------------|------------------------------
ccs811 basic version with nWAKE and nRESET handling | 820 byte
ccs811 basic version w/o nWAKE and nRESET handling | 970 byte
ccs811 full version | 1426 byte

Reply:
Regarding nWAKE, we must take a decision between code size and energy consumption. To me they are equally important. So I think it wouldn't make sense to have to pull in the full version of the driver in order to ensure a minimal energy consumption. Having 100 extra bytes seems reasonable.

I also thought that some code optimization could still be done by excluding some params attributes if not using the full driver version (not tested but I'll have a look and comment inline after)

Reply:
@gschorcht, see https://github.com/gschorcht/RIOT-Xtensa-ESP/pull/2, with the basic version using nWAKE and nRESET, I get a 894 bytes code size for ccs811.a. The full version is 1422 bytes.

Reply:
I merged it. Thanks. Now we have nWAKE and nRESET in base version.